Setup the IH assistant (#46)
DataDog/dd-sdk-ios-apollo-interceptor: tools/issue_handler/README.md

# GitHub Issue Handler

> A tool that automatically analyzes new GitHub issues for the Datadog iOS Apollo Interceptor using OpenAI and posts summaries to Slack.

Reviewer: Maybe slightly clearer? Defer to you

```suggestion
A tool for the Datadog iOS Apollo Interceptor repository that analyzes new GitHub issues using OpenAI and posts summaries to Slack.
```

## Features

- 🔍 Fetches GitHub issue details
- 🤖 Analyzes issues using OpenAI
- 💬 Posts analysis to Slack
- 🔄 Runs automatically on new issues via GitHub Actions
- 🛠️ Can be run manually for specific issues

## Setup

### 1. Create Virtual Environment

```bash
# Create a virtual environment
python3 -m venv venv

# Activate the virtual environment
source venv/bin/activate
```

### 2. Install Dependencies

```bash
pip install -r requirements.txt
```

### 3. Configure Environment

> First, create your local environment file:

Reviewer: ```suggestion
Create your local environment file:
```

```bash
# Create .env file from template
./setup_env.sh
```

> This creates a `.env` file that you will need to fill with the required tokens. The file looks like the following:

Reviewer: ```suggestion
This creates a `.env` file like the following:
```

```bash
# GitHub token with repo access
GITHUB_TOKEN=

# OpenAI API token
OPENAI_TOKEN=

# Prompt for OpenAI to analyze GitHub issues
OPENAI_SYSTEM_PROMPT=

# Slack webhook URL (for posting notifications)
SLACK_WEBHOOK_URL=

# Slack channel ID (where to post notifications)
SLACK_CHANNEL_ID=

# Github repository (optional, defaults to DataDog/dd-sdk-ios-apollo-interceptor)
GITHUB_REPOSITORY=
```

> ### 4. Required Tokens

Reviewer: Using active tense to match the other steps, and using "variables" since not everything here is a token.

```suggestion
### 4. Set Variables
```

> You need several tokens to use the tool locally. See the instructions below to get each token:

Reviewer: ```suggestion
Complete the following steps to retrieve and set the required values in your `.env` file.
```

#### GitHub Token
1. Go to [GitHub Settings > Developer Settings > Personal Access Tokens](https://github.com/settings/tokens)
2. Generate a new token with `repo` scope
3. Copy the token

#### OpenAI Token
1. Go to [OpenAI API Keys](https://platform.openai.com/api-keys)
2. Create a new API key
3. Copy the key (starts with `sk-`)

#### OpenAI System Prompt
1. Create your AI prompt for analyzing GitHub issues
2. This prompt is used to guide the AI's analysis and response generation

#### Slack Access
The iOS SDK team already has a Slack app configured for this tool. You need:

1. **Slack Webhook URL**:

>    - Create a webhook in your Slack workspace: https://api.slack.com/apps

Reviewer: The next step is where you create a new webhook, so cutting that out here

```suggestion
   - Navigate to https://api.slack.com/apps
```

   - Go to "Incoming Webhooks", and create a new webhook
   - Copy the webhook URL (starts with `https://hooks.slack.com/services/`)
   - Store it securely, and never commit it

2. **Slack Channel Setup**:
   - Get the channel ID by clicking the channel's details in Slack, and copy the ID at the bottom

>    - The ID starts with "C" (for example, "C12345678")

Reviewer: ```suggestion
   - Get the channel ID by clicking the channel's details in Slack, and copy the ID at the bottom
     - The ID starts with "C" (for example, "C12345678")
```

   - Invite the bot to your channel: `/invite @bot-name`

Note: If you need to post to a new channel, make sure to:
1. Invite the bot to that channel first
2. Use that channel's ID in your `.env` file

## Usage

### Run Manually

> Make sure your virtual environment is activated, then analyze a specific issue:

Reviewer: ```suggestion
Activate your virtual environment, then analyze a specific issue:
```

```bash
python src/analyze_issue.py ISSUE_NUMBER
```

For example:
```bash
python src/analyze_issue.py 1234
```

### GitHub Action

The tool runs:
1. Automatically when a new issue is opened in the repository
2. Manually when triggered through GitHub Actions with an issue number

The GitHub Action uses repository secrets for authentication. These are already configured in the repository settings:
- `GITHUB_TOKEN` (automatically provided)
- `OPENAI_TOKEN`
- `OPENAI_SYSTEM_PROMPT`
- `SLACK_WEBHOOK_URL`
- `SLACK_CHANNEL_ID`
- `GITHUB_REPOSITORY` (defaults to `DataDog/dd-sdk-ios-apollo-interceptor`)

## Output

For each issue, the tool does the following:
1. Analyze the issue using OpenAI

> 2. Post a message to Slack containing:

Reviewer: ```suggestion
1. Analyzes the issue using OpenAI
2. Posts a message to Slack containing:
```

   - GitHub issue notification
   - Analysis summary
   - Suggested response
   - Confidence level

## Development

### Project structure

- Source code is in `src/`
- Tests are in `tests/`
- Environment variables are managed via `.env`

### Architecture

Main Components:
- analyze_issue.py - Main entry point that orchestrates the workflow
- github_handler.py - Fetches GitHub issue details via API
- openai_handler.py - Analyzes issues using OpenAI

> - slack_handler.py - Posts notifications and analysis to Slack

Reviewer: ```suggestion
- `analyze_issue.py` - Main entry point that orchestrates the workflow
- `github_handler.py` - Fetches GitHub issue details via API
- `openai_handler.py` - Analyzes issues using OpenAI
- `slack_handler.py` - Posts notifications and analysis to Slack
```

### Workflow 

- GitHub issue is opened → triggers GitHub Action
- Fetches issue details from GitHub API
- Analyzes issue with OpenAI using a custom prompt
- Posts Github issue notification and analysis on Slack

## Security Notes

- Never commit your `.env` file
- Keep all tokens secure and private
- Use GitHub Secrets for CI/CD
- The `.env` file is git-ignored by default

## Running Tests

> Make sure your virtual environment is activated, then run all tests:

Reviewer: ```suggestion
Activate your virtual environment, then run all tests:
```
